test/e2e-common.sh

```diff
@@ -210,6 +211,85 @@ function enable_monitoring(){
       --role roles/cloudtrace.agent
 }
 
+# test_authentication_check_for_brokercell tests the authentication check function for BrokerCell.
+# This test required the entire BrokerCell to be non-functional.
+# In order to avoid running in parallel with other tests, we put it in the shell script, rather than more common go tests.
+# TODO Once we support multiple BrokerCells with distinct authentication mechanisms, make this a normal test that runs in parallel with the others.
+function test_authentication_check_for_brokercell() {
+  echo "Starting authentication check test for brokercell."
+  local auth_mode=${1}
+  kubectl apply -f ${CONFIG_WARMUP_GCP_BROKER}
+
+  echo "Starting authentication check test which is running outside of the broker related Pods."
+  wait_until_brokercell_authentication_check_pending "$(non_pod_check_keywords)" || return 1
+
+  echo "Starting authentication check test which is running inside of the broker related Pods."
+  apply_invalid_auth "$auth_mode" || return 1
+  wait_until_brokercell_authentication_check_pending "$(pod_check_keywords "$auth_mode")" || return 1
+
+  # Clean up all the testing resources.
+  echo "Authentication check test finished, waiting until all broker related testing resources deleted."
+  delete_invalid_auth "$auth_mode" || return 1
+  kubectl delete -f ${CONFIG_WARMUP_GCP_BROKER}
+  kubectl delete brokercell default -n "${CONTROL_PLANE_NAMESPACE}"
+  kubectl wait pod --for=delete -n "${CONTROL_PLANE_NAMESPACE}" --selector=brokerCell=default --timeout=5m || return 1
+}
+
+function wait_until_brokercell_authentication_check_pending() {
+  local keywords=${1}
+  echo "Waiting until brokercell authentication check pending."
+  for i in {1..150}; do #timeout after 5 minutes
+    local message=$(kubectl get brokercell default -n "${CONTROL_PLANE_NAMESPACE}" -o=jsonpath='{.status.conditions[?(@.reason == "AuthenticationCheckPending")].message}' --ignore-not-found=true)
+    if [[ "$message" == *"$keywords"* ]]; then
+      return 0
+    fi
+    sleep 2
+  done
+  echo -e "\n\nERROR: timeout waiting for brokercell authentication check pending with correct key message: ${keywords}"
+  echo -e "brokercell object YAML: $(kubectl get brokercell default -n "${CONTROL_PLANE_NAMESPACE}" -o yaml)"
+  return 1
+}
+
+function pod_check_keywords() {
+  local auth_mode=${1}
+  if [ "${auth_mode}" == "secret" ]; then
+    echo "error getting the token, probably due to the key stored in the Kubernetes Secret is expired or revoked"
+  elif [ "${auth_mode}" == "workload_identity" ]; then
+    echo "the Pod is not fully authenticated, probably due to corresponding k8s service account and google service account do not establish a correct relationship"
```

Comment:
nits: I don't think you need to have the full message, it is hard to maintain if we later change the message. I suggest to just keep the important info, like `error getting the token`.

Reply:
This is not a full message. The full message is even longer... I would like to keep this message, to clearly test out which step it goes wrong.

Reply:
I see. As long as we only catch important errors you think necessary, it lgtm.